iothub/device/src/InternalClient.cs

```diff
@@ -1273,6 +1290,130 @@ public Task<Message> ReceiveAsync(CancellationToken cancellationToken)
+                throw;
+            }
+            finally
+            {
+                try
+                {
+                    _deviceReceiveMessageSemaphore.Release();
+                }
+                catch (SemaphoreFullException)
+                {
+                    // this semaphore is typically grabbed at the start of the method, but if
+                    // this method is canceled while waiting to grab the semaphore, then this semaphore.release
+                    // will throw this SemaphoreFullException since it was never grabbed in the first place
+                    if (Logging.IsEnabled)
+                    {
+                        Logging.Info(this, "SemaphoreFullException thrown while releasing" +
+                        " message receiving semaphore, but will be ignored since that means the semaphore" +
+                        " is available for other threads to grab again anyways");
+                    }
+                }
+
+                if (Logging.IsEnabled)
+                {
+                    Logging.Exit(this, messageHandler, userContext, nameof(SetReceiveMessageHandlerAsync));
+                }
+            }
+        }
+
+        // Enable telemetry downlink for devices
+        private Task EnableReceiveMessageAsync(CancellationToken cancellationToken)
+        {
+            // The telemetry downlink needs to be enabled only for the first time that the _receiveMessageCallback delegate is set.
+            return _deviceReceiveMessageCallback == null
+                ? InnerHandler.EnableReceiveMessageAsync(cancellationToken)
+                : TaskHelpers.CompletedTask;
+        }
+
+        // Disable telemetry downlink for devices
+        private Task DisableReceiveMessageAsync(CancellationToken cancellationToken)
+        {
+            // The telemetry downlink should be disabled only after _receiveMessageCallback delegate has been removed.
+            return _deviceReceiveMessageCallback == null
+                ? InnerHandler.DisableReceiveMessageAsync(cancellationToken)
+                : TaskHelpers.CompletedTask;
+        }
+
+        /// <summary>
+        /// The delegate for handling c2d messages received
+        /// <param name="message">The message received</param>
+        /// </summary>
+        private async Task OnDeviceMessageReceivedAsync(Message message)
+        {
+            if (Logging.IsEnabled)
+            {
+                Logging.Enter(this, message, nameof(OnDeviceMessageReceivedAsync));
+            }
+
+            if (message == null)
+            {
+                return;
+            }
+
+            Tuple<ReceiveMessageCallback, object> callbackContextTuple = null;
+
+            try
+            {
+                // Wait to acquire the _deviceReceiveMessageSemaphore. This ensures that we get a reference to the latest callback set on the device client.
+                await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
```

Comment:
why not use lock?

Reply:
I don't have a really strong argument for this, other than semaphores being easier to track and maintain, hence my preference.